### PR TITLE
OSAC 164 - [Cryptography] Insecure unauthenticated encryption in Encrypt Text

### PR DESCRIPTION
**Pull request by alcxs**
Description
Encrypt Text - is using an antiquated mode of AES encryption, one that does not authenticate the message
How to test
load Credential package and use AES GCM algorithm
Jira
[OSAC-164](https://uipath.atlassian.net/browse/OSAC-164)
Label:
24S

### CODE REVIEW

File: Activities/Cryptography/UiPath.Cryptography/CryptographyHelper.cs [hunk 1]
```diff
@@ -45,74 +54,117 @@ public static byte[] HashDataWithKey(KeyedHashAlgorithms keyedHashAlgorithm, byt
...
             {
-                byte[] encrypted;
-                byte[] salt = new byte[PBKDF2_SaltSizeBytes];
-                int maxKeySize = GetLegalKeySizes(algorithm).Max();
-
-                _rng.GetBytes(salt);
-                using (Rfc2898DeriveBytes pbkdf2 = new Rfc2898DeriveBytes(key, salt, PBKDF2_Iterations))
+                return EncryptAesGcmNet461(inputBytes, key, out result);
+            }
+            else
```
Comment (ovidiuiliescu) [inline Activities/Cryptography/UiPath.Cryptography/CryptographyHelper.cs]: Please refactor this to be more readable, as the condition itself is the same in both cases. E.g.

```
if (symmetricAlgorithm == SymmetricAlgorithms.AESGCM)
{
#if NET5_0
...
#else
...
#endif
}
```

Same elsewhere in the code (e.g in the decrypt methods).

Please also use don't explicitly check for NET5_0 and NET461, as the targets may change later and we won't go back through the code to replace them, resulting in potential issues.

Instead, use `NETCOREAPP` and `NETFRAMEWORK` as they are more flexible, see the table at the end of https://docs.microsoft.com/en-us/dotnet/standard/frameworks#how-to-specify-a-target-framework .

File: Activities/Cryptography/UiPath.Cryptography/CryptographyHelper.cs [hunk 2]
```diff
@@ -45,74 +54,117 @@ public static byte[] HashDataWithKey(KeyedHashAlgorithms keyedHashAlgorithm, byt
...
-                Buffer.BlockCopy(algorithm.IV, 0, result, salt.Length, algorithm.IV.Length);
-                Buffer.BlockCopy(encrypted, 0, result, salt.Length + algorithm.IV.Length, encrypted.Length);
+                return result;
+#if NET461
             }
-
-            return result;
+#endif
+#if NET5_0
+            }
```
Comment (ovidiuiliescu) [inline Activities/Cryptography/UiPath.Cryptography/CryptographyHelper.cs:96]: This kind of preprocessing is really bad practice and is not needed. Having a single closing bracket as the start of a preprocessing region is a strong code smell.

In this particular case, you don't need to have the code like this:

```
if (...) return A
else if (...) return B
else return C
```

You can just have

```
if (...) return A

if (...) return B

return C
```

It makes the code much more readable when preprocessing directives are involved.

File: Activities/Cryptography/UiPath.Cryptography/CryptographyHelper.cs [hunk 3]
```diff
@@ -260,5 +315,145 @@ private static int[] GetLegalKeySizes(SymmetricAlgorithm algorithm)
...
+            using (Rfc2898DeriveBytes pbkdf2 = new Rfc2898DeriveBytes(key, salt, PBKDF2_Iterations))
+            {
+                var Key = pbkdf2.GetBytes(32); //256 bit key
+                decrypted = DecryptWithBouncyCastle(encryptedData, iv, tag, Key);
+            }
+
+            return decrypted;
+        }
+
+        private static byte[] EncryptWithBouncyCastle(byte[] plaintext, byte[] key, byte[] nonce, out byte[] tag)
```
Comment (ovidiuiliescu) [inline Activities/Cryptography/UiPath.Cryptography/CryptographyHelper.cs]: Let's rename this to `EncryptAesGcmWithBouncyCastle` or similar, so that future us will know what algorithm the code uses.

Same for decrypt.

File: Activities/Database/Database.build.props [hunk 4]
```diff
@@ -2,6 +2,6 @@
   <PropertyGroup>
     <Company>UiPath</Company>
     <Copyright>© UiPath</Copyright>
-    <VersionPrefix>1.5.0</VersionPrefix>
+    <VersionPrefix>1.6.0</VersionPrefix>
```
Comment (ovidiuiliescu) [inline Activities/Database/Database.build.props]: Why are these changed here ?
Same for other package versions.

**Review by ovidiuiliescu — CHANGES_REQUESTED** [review]

**Review by ovidiuiliescu — COMMENTED** [review]
We also need some unit tests around this new functionality.

Ideally cross-check .NET 5 with BouncyCastle and vice-versa.

File: Activities/Cryptography/UiPath.Cryptography/CryptographyHelper.cs [hunk 5]
```diff
@@ -260,5 +315,145 @@ private static int[] GetLegalKeySizes(SymmetricAlgorithm algorithm)
...
+
+            result = new byte[salt.Length + algorithmIV.Length + encrypted.Length + tag.Length];
+            Buffer.BlockCopy(salt, 0, result, 0, salt.Length);
+            Buffer.BlockCopy(algorithmIV, 0, result, salt.Length, algorithmIV.Length);
+            Buffer.BlockCopy(encrypted, 0, result, salt.Length + algorithmIV.Length, encrypted.Length);
+            Buffer.BlockCopy(tag, 0, result, salt.Length + algorithmIV.Length + encrypted.Length, tag.Length);
+
+            return result;
+        }
+        private static byte[] DecryptAesGcmNet5(byte[] inputBytes, byte[] key, out byte[] decrypted)
```
Comment (ovidiuiliescu) [inline Activities/Cryptography/UiPath.Cryptography/CryptographyHelper.cs]: Nitpick: please insert an empty line before this method.

File: Activities/Cryptography/UiPath.Cryptography/CryptographyHelper.cs [hunk 6]
```diff
@@ -222,7 +274,10 @@ private static SymmetricAlgorithm GetSymmetricAlgorithm(SymmetricAlgorithms symm
             switch (symmetricAlgorithm)
             {
                 case SymmetricAlgorithms.AES:
-                    return new AesCryptoServiceProvider(); // TODO: Use AesCng after upgrading to .NET Framework 4.6.2
+                    return new AesCryptoServiceProvider(); // kept for backwords compat
+                case SymmetricAlgorithms.AESGCM:
```
Comment (ovidiuiliescu) [inline Activities/Cryptography/UiPath.Cryptography/CryptographyHelper.cs:260]: Nitpick: Empty line before this

Also before the default below.

File: Activities/Cryptography/UiPath.Cryptography/CryptographyHelper.cs [hunk 7]
```diff
@@ -260,5 +315,145 @@ private static int[] GetLegalKeySizes(SymmetricAlgorithm algorithm)
 
             return keySizes.ToArray();
         }
+
+#if NET5_0
+        private static byte[] EncryptAesGcmNet5(byte[] inputBytes, byte[] key, out byte[] result)
```
Comment (ovidiuiliescu) [inline Activities/Cryptography/UiPath.Cryptography/CryptographyHelper.cs]: There's a lot of overlap between this method and the 461 one.

Same for the decryption counterparts.

Can we extract the common parts into dedicated methods and reuse them?

Or, since it seems that only the part inside the `using`s is different, maybe create some helper methods around those and have a single set of Encrypt/Decrypt methods, which call the appropriate helper methods inside them via `#ifdef` ?

Something like

```
private static byte[] EncryptAesGcm(...)
{
// Init

using()
{
#if NET5_0
EncryptAesGcmWith50(...);
#endif

#if NET461
EncryptAesGcmWith461(...);
#endif
}

// Return result
}
```

This also has the advantage of removing a lot of the preprocessing directives elsewhere, since we have a common entry point method for both NET 5.0 and NET 461, and we only need preprocessing inside that method.

Either way, we need to cut down on the code duplication.

**Review by ovidiuiliescu — CHANGES_REQUESTED** [review]

File: Activities/Cryptography/UiPath.Cryptography/CryptographyHelper.cs (hunk 1 above)
Comment (alcxs) [reply Activities/Cryptography/UiPath.Cryptography/CryptographyHelper.cs]: changed to match recommendation

File: Activities/Cryptography/UiPath.Cryptography/CryptographyHelper.cs (hunk 2 above)
Comment (alcxs) [reply Activities/Cryptography/UiPath.Cryptography/CryptographyHelper.cs:96]: changed, as requested

File: Activities/Database/Database.build.props (hunk 4 above)
Comment (alcxs) [reply Activities/Database/Database.build.props]: renamed

File: Activities/Cryptography/UiPath.Cryptography/CryptographyHelper.cs (hunk 5 above)
Comment (alcxs) [reply Activities/Cryptography/UiPath.Cryptography/CryptographyHelper.cs]: added

File: Activities/Cryptography/UiPath.Cryptography/CryptographyHelper.cs (hunk 6 above)
Comment (alcxs) [reply Activities/Cryptography/UiPath.Cryptography/CryptographyHelper.cs:260]: added

File: Activities/Cryptography/UiPath.Cryptography/CryptographyHelper.cs (hunk 7 above)
Comment (alcxs) [reply Activities/Cryptography/UiPath.Cryptography/CryptographyHelper.cs]: modified

File: Activities/Cryptography/UiPath.Cryptography/CryptographyHelper.cs (hunk 3 above)
Comment (alcxs) [reply Activities/Cryptography/UiPath.Cryptography/CryptographyHelper.cs]: renamed

File: Activities/Cryptography/UiPath.Cryptography/CryptographyHelper.cs [hunk 8]
```diff
@@ -46,68 +55,96 @@ public static byte[] EncryptData(SymmetricAlgorithms symmetricAlgorithm, byte[]
...
-                byte[] salt = new byte[PBKDF2_SaltSizeBytes];
-                int maxKeySize = GetLegalKeySizes(algorithm).Max();
+#if NETCOREAPP
 
-                _rng.GetBytes(salt);
-                using (Rfc2898DeriveBytes pbkdf2 = new Rfc2898DeriveBytes(key, salt, PBKDF2_Iterations))
-                {
-                    algorithm.Key = pbkdf2.GetBytes(maxKeySize);
-                }
+                return EncryptAesGcm(inputBytes, key, out result);
```
Comment (ovidiuiliescu) [inline Activities/Cryptography/UiPath.Cryptography/CryptographyHelper.cs]: Why is the same method called in both branches?

If we call the same code in both cases, let's remove the ifdefs.

Same for the DecryptData method.

**Review by ovidiuiliescu — CHANGES_REQUESTED** [review]

File: Activities/Cryptography/UiPath.Cryptography/CryptographyHelper.cs (hunk 8 above)
Comment (alcxs) [reply Activities/Cryptography/UiPath.Cryptography/CryptographyHelper.cs]: sure, removed.

File: Activities/Cryptography/UiPath.Cryptography/CryptographyHelper.cs [hunk 9]
```diff
@@ -260,5 +298,124 @@ private static int[] GetLegalKeySizes(SymmetricAlgorithm algorithm)
 
             return keySizes.ToArray();
         }
+
+        private static byte[] EncryptAesGcm(byte[] inputBytes, byte[] key, out byte[] result)
```
Comment (ovidiuiliescu) [inline Activities/Cryptography/UiPath.Cryptography/CryptographyHelper.cs]: Any good reason why we have result as an out argument and also as a return value?
Can we have it just as a return value?

Same for decrypt method.

Comment (alcxs) [reply Activities/Cryptography/UiPath.Cryptography/CryptographyHelper.cs]: removed as out parameter
